rust/src/bwrap.rs

```diff
@@ -0,0 +1,481 @@
+            "--chdir",
+            "/",
+            "--die-with-parent", /* Since 0.1.8 */
+            /* Here we do all namespaces except the user one.
+             * Down the line we want to do a userns too I think,
+             * but it may need some mapping work.
+             */
+            "--unshare-pid",
+            "--unshare-uts",
+            "--unshare-ipc",
+            "--unshare-cgroup-try",
+        ];
+
+        for d in RO_BINDS {
+            argv.push("--ro-bind");
+            argv.push(d);
+            argv.push(d);
+        }
+
+        if !running_in_nspawn() {
+            argv.push("--unshare-net");
+        }
+
+        /* Capabilities; this is a subset of the Docker (1.13 at least) default.
+         * Specifically we strip out in addition to that:
+         *
+         * "cap_net_raw" (no use for this in %post, and major source of security vulnerabilities)
+         * "cap_mknod" (%post should not be making devices, it wouldn't be persistent anyways)
+         * "cap_audit_write" (we shouldn't be auditing anything from here)
+         * "cap_net_bind_service" (nothing should be doing IP networking at all)
+         *
+         * But crucially we're dropping a lot of other capabilities like
+         * "cap_sys_admin", "cap_sys_module", etc that Docker also drops by default.
+         * We don't want RPM scripts to be doing any of that. Instead, do it from
+         * systemd unit files.
+         *
+         * Also this way we drop out any new capabilities that appear.
+         */
+        if nix::unistd::getuid() == nix::unistd::Uid::from_raw(0) {
+            argv.extend(&["--cap-drop", "ALL"]);
+            for cap in ADDED_CAPABILITIES {
+                argv.push("--cap-add");
+                argv.push(cap);
+            }
+        }
+
+        let mut argv: Vec<_> = argv.into_iter().map(|s| s.to_string()).collect();
+
+        for &name in USR_LINKS.iter() {
+            let stbuf = rootfs_fd.metadata_optional(name)?;
+            if let Some(stbuf) = stbuf {
+                if !matches!(stbuf.simple_type(), openat::SimpleType::Symlink) {
+                    continue;
+                }
+
+                argv.push("--symlink".to_string());
+                argv.push(format!("usr/{}", name));
+                argv.push(format!("/{}", name));
+            }
+        }
+
+        Ok(Self {
+            rootfs_fd,
+            executed: false,
+            argv,
+            launcher,
+            child_argv0: None,
+            rofiles_mounts: Vec::new(),
+        })
+    }
+
+    fn setup_rofiles(&mut self, path: &str) -> Result<()> {
+        let mnt = RoFilesMount::new(&self.rootfs_fd, path)?;
+        let tmpdir_path = mnt.path().to_str().expect("tempdir str");
+        self.bind_readwrite(tmpdir_path, path);
+        self.rofiles_mounts.push(mnt);
+        Ok(())
+    }
+
+    /// Create a new bwrap instance with the provided level of mutability.
+    pub(crate) fn new_with_mutability(
```

Comment:
Minor: would be cleaner to reorder this one to be right after the implementation of `new()` so the constructors are together.